File: mobile/lib/protocol/auth_payload.dart
import 'dart:typed_data';

import 'frame.dart';
import 'payload_codec.dart';

const int authCredentialTypeToken = 0x01;
const int authCredentialTypePairingCode = 0x02;

class AuthPayload {
  const AuthPayload({
    required this.credentialType,
    required this.credential,
    required this.deviceId,
    required this.challengeResponse,
  });

  final int credentialType;
  final String credential;
  final String deviceId;
  final Uint8List challengeResponse;

  @override
  bool operator ==(Object other) =>
      other is AuthPayload &&
      credentialType == other.credentialType &&
      credential == other.credential &&
      deviceId == other.deviceId &&
      _bytesEqual(challengeResponse, other.challengeResponse);

  @override
  int get hashCode => Object.hash(
        credentialType,
        credential,
        deviceId,
        Object.hashAll(challengeResponse),
      );
}

class AuthPayloadCodec {
  static Uint8List encode(AuthPayload payload) {
    if (!_isKnownCredentialType(payload.credentialType)) {
      throw const ProtocolException('Unknown credentialType.');
    }
    if (payload.credentialType == authCredentialTypeToken &&
        payload.credential.isNotEmpty) {
      throw const ProtocolException('Token credential must be empty.');
    }
    if (payload.challengeResponse.length != 32) {
      throw const ProtocolException(
          'challengeResponse must be exactly 32 bytes.');
    }

    final writer = PayloadWriter()
      ..writeUInt8(payload.credentialType)
      ..writeString(payload.credential)
      ..writeString(payload.deviceId, minLength: 1, maxLength: 64)
      ..writeBytes(payload.challengeResponse);
    return writer.toBytes();
  }

  static AuthPayload decode(List<int> bytes) {
    final reader = PayloadReader(bytes);
    final credentialType = reader.readUInt8();
    final credential = reader.readString();
    final deviceId = reader.readString(minLength: 1, maxLength: 64);
    final challengeResponse = reader.readBytes(32);
    reader.expectEnd();

    if (!_isKnownCredentialType(credentialType)) {
      throw const ProtocolException('Unknown credentialType.');
    }
    if (credentialType == authCredentialTypeToken && credential.isNotEmpty) {
      throw const ProtocolException('Token credential must have length 0.');
    }

    return AuthPayload(
      credentialType: credentialType,
      credential: credential,
      deviceId: deviceId,
      challengeResponse: challengeResponse,
    );
  }

  static bool _isKnownCredentialType(int credentialType) =>
      credentialType == authCredentialTypeToken ||
      credentialType == authCredentialTypePairingCode;
}

bool _bytesEqual(Uint8List a, Uint8List b) {
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i++) {
    if (a[i] != b[i]) return false;
  }
  return true;
}

File: desktop/Protocol/AuthPayload.cs
namespace CTRL.Desktop.Protocol;

public sealed record AuthPayload(
    byte CredentialType,
    string Credential,
    string DeviceId,
    byte[] ChallengeResponse);

public static class AuthPayloadCodec
{
    public const byte CredentialTypeToken = 0x01;
    public const byte CredentialTypePairingCode = 0x02;

    public static byte[] Encode(AuthPayload payload)
    {
        if (!IsKnownCredentialType(payload.CredentialType))
            throw new ProtocolException("Unknown credentialType.");
        if (payload.CredentialType == CredentialTypeToken &&
            !string.IsNullOrEmpty(payload.Credential))
            throw new ProtocolException("Token credential must be empty.");
        if (payload.ChallengeResponse is null || payload.ChallengeResponse.Length != 32)
            throw new ProtocolException("challengeResponse must be exactly 32 bytes.");

        var writer = new PayloadWriter();
        writer.WriteUInt8(payload.CredentialType);
        writer.WriteString(payload.Credential);
        writer.WriteString(payload.DeviceId, 1, 64);
        writer.WriteBytes(payload.ChallengeResponse);
        return writer.ToArray();
    }

    public static AuthPayload Decode(byte[] payload)
    {
        var reader = new PayloadReader(payload);
        var credentialType = reader.ReadUInt8();
        var credential = reader.ReadString();
        var deviceId = reader.ReadString(1, 64);
        var challengeResponse = reader.ReadBytes(32);
        reader.ExpectEnd();

        if (!IsKnownCredentialType(credentialType))
            throw new ProtocolException("Unknown credentialType.");
        if (credentialType == CredentialTypeToken && credential.Length != 0)
            throw new ProtocolException("Token credential must have length 0.");

        return new AuthPayload(credentialType, credential, deviceId, challengeResponse);
    }

    private static bool IsKnownCredentialType(byte credentialType) =>
        credentialType == CredentialTypeToken ||
        credentialType == CredentialTypePairingCode;
}
